executor/join.go

```diff
@@ -665,76 +658,106 @@ func (e *HashJoinExec) handleFetchInnerAndBuildHashTablePanic(r interface{}) {
 }
 
 func (e *HashJoinExec) fetchInnerAndBuildHashTable(ctx context.Context) {
-	// innerResultCh transfers inner chunk from inner fetch to build hash table.
-	innerResultCh := make(chan *chunk.Chunk, 1)
-	doneCh := make(chan struct{})
-	// TODO: just used in test now, thus we return directly in this if-branch.
-	// Remained work:
-	// 1. parallel build hash tables for the inner sub-partitions
-	// 2. partition the outer relation into sub-partitions
-	// 3. parallel join corresponded inner-partitions with outer-partitions
-	if e.ctx.GetSessionVars().EnableRadixJoin {
-		go util.WithRecovery(func() {
-			e.fetchInnerRows(ctx, innerResultCh, doneCh)
-			if !e.evalRadixBit() {
-				return
-			}
-			if e.partitionInnerRows() != nil {
-				return
-			}
+	// Fetch all the data of the inner relation into memory.
+	if err := e.fetchInnerRows(ctx); err != nil {
+		e.innerFinished <- err
+		return
+	}
 
-		}, nil)
+	// Do parallel partition if needed.
+	needPartition := e.ctx.GetSessionVars().EnableRadixJoin && e.evalRadixBit()
+
+	// non-partitioned hash join
+	if !needPartition {
+		if err := e.buildGlobalHashTable(); err != nil {
+			e.innerFinished <- err
+		}
 		return
 	}
 
-	go util.WithRecovery(func() { e.fetchInnerRows(ctx, innerResultCh, doneCh) }, nil)
+	// radix-partitioned hash join
+	if err := e.partitionInnerRows(); err != nil {
+		e.innerFinished <- err
+		return
+	}
+	if err := e.buildHashTable4Partitions(); err != nil {
+		e.innerFinished <- err
+	}
+}
 
-	// TODO: Parallel build hash table. Currently not support because `mvmap` is not thread-safe.
-	err := e.buildHashTableForList(innerResultCh)
-	if err != nil {
-		e.innerFinished <- errors.Trace(err)
-		close(doneCh)
+func (e *HashJoinExec) wait4BuildHashTable(wg *sync.WaitGroup, finishedCh chan error) {
+	wg.Wait()
+	close(finishedCh)
+}
+
+func (e *HashJoinExec) buildHashTable4Partitions() error {
+	e.hashTables = make([]*mvmap.MVMap, len(e.innerParts))
+	buildFinishedCh := make(chan error, e.concurrency)
+	wg := &sync.WaitGroup{}
+	wg.Add(int(e.concurrency))
+	go e.wait4BuildHashTable(wg, buildFinishedCh)
+	for i := 0; i < int(e.concurrency); i++ {
+		workerID := i
+		go util.WithRecovery(func() {
+			defer wg.Done()
+			e.doBuild(workerID, buildFinishedCh)
+		}, nil)
 	}
-	// wait fetchInnerRows be finished.
-	for range innerResultCh {
+	return <-buildFinishedCh
+}
+
+func (e *HashJoinExec) doBuild(workerID int, finishedCh chan error) {
+	var err error
+	keyBuf, valBuf := make([]byte, 0, 64), make([]byte, 4)
+	for i := workerID; i < len(e.innerParts); i += int(e.concurrency) {
+		if e.innerParts[i] == nil {
+			continue
+		}
+		e.hashTables[i] = mvmap.NewMVMap()
+		keyBuf = keyBuf[:0]
+		for rowIdx, numRows := 0, e.innerParts[i].NumRows(); rowIdx < numRows; rowIdx++ {
+			// Join-key can be promised to be not null in a partition, so we do not check it.
+			_, keyBuf, err = e.getJoinKeyFromChkRow(false, e.innerParts[i].GetRow(rowIdx), keyBuf)
```

Comment:
Should we skip this inner row if there is a `null` in the join key as what we did in `buildGlobalHashTable()`?

Another thing is, we are going to make the hash join to be NULL-Aware, that is, to produce the correct result for "left outer semi join" or "anti left outer semi join". We should use a global variable to record whether there is a `NULL` value in the join key of inner side? It's fine for this PR to not consider the NULL-Awareness, we can do it in the future PRs. @eurekaka May take a look at this place in the future.

Reply:
Inner rows with null-join-key(s) were filter during [partition phase](https://github.com/pingcap/tidb/blob/master/executor/join.go#L358-L359).

Reply:
OK, could you add a comment in here to help understanding the code in the future?